[CMake] Integrate vcpkg into Windows build

donny-dont opened the pull request:
#### 32342c72df2d86f7d2cb5ab8b38d3b4289c225f4
<pre>
[CMake] Integrate vcpkg into Windows build
<a href="https://bugs.webkit.org/show_bug.cgi?id=282800">https://bugs.webkit.org/show_bug.cgi?id=282800</a>

Reviewed by Ross Kirsling.

Use `vcpkg` to handle the third party requirements of the Windows WebKit build.
The requirements will be built locally within the build and no longer
distributed from WebKitForWindows/WebKitRequirements. While all port definitions
are in that repository a migration to the canonical vcpkg registry at
Microsoft/vcpkg will be prioritized.

Removes the code that would download prebuilt binaries to `WebKitLibraries/win`.

This change will be used to support additional libraries in the future.

* .gitignore:
* Source/cmake/OptionsJSCOnly.cmake:
* Source/cmake/OptionsWin.cmake:
* Tools/Scripts/build-webkit:
* Tools/Scripts/download-github-release.py: Removed.
* Tools/Scripts/update-webkit-win-libs.py:
* Tools/Scripts/webkitdirs.pm:
(shouldUseVcpkg):
(generateBuildSystemFromCMakeProject):
(vcpkgArgsFromFeatures):
(windowsLibrariesDir): Deleted.
* Tools/Scripts/webkitperl/BuildSubproject.pm:
* WebKitLibraries/triplets/x64-windows-webkit.cmake: Added.
* vcpkg-configuration.json: Added.
* vcpkg.json: Added.

Canonical link: <a href="https://commits.webkit.org/295042@main">https://commits.webkit.org/295042@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9237033312bfc38be24e0380ebe58cde6c07590c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/103384 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/23066 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/13383 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/108558 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/54027 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/105423 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/23475 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/31617 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/78562 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/35508 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/106390 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/18191 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/93264 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/58896 "Passed tests") | 
| | [💥 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/18015 "An unexpected error occured. Recent messages:Printed configuration") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/11308 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/53384 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/96059 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/87810 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/11369 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/110934 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/101995 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/30531 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/22528 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/87560 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/30892 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/89462 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/87198 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22299 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/32062 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/9786 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/24858 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/30459 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/125628 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/30259 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/34828 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/33590 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/31820 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

fujii reviewed (changes requested):
I don't like this approach.
* Buildbot workers have to compile all required libraries. But, we have no enough VM resources.
* It increases the built artifact size if all compiled DLL are copies to `bin` directory. (layout tests bots need built DLLs)

I think current WebKitRequirements64.zip approach is better.

fujii commented on Tools/Scripts/update-webkit-win-libs.py:
Don't remove this file. You can remove this file after deploying EWS and post-commit Buildbot.
And, I think we should keep this file with a content invoking "vcpkg install" for layout test buildbot rather than including all dependent DLL to a zip file.

fujii reviewed (changes requested):
Why don't you use GitHub packages binary cache?
https://learn.microsoft.com/en-us/vcpkg/reference/binarycaching

donny-dont commented on vcpkg.json:
```diff
@@ -0,0 +1,92 @@
+    "icu"
+  ],
+  "features": {
+    "avif": {
+      "description": "Build with AVIF image support",
+      "dependencies": [
+        {
+          "name": "libavif",
+          "features": [
+            "aom"
+          ]
+        }
+      ]
+    },
```
@fujii @tats-u with `vcpkg` you can add features for the build. In this case it'll use the upstream `vcpkg` registry to build with avif support.

The `build-webkit` script would need to have knowledge of this going forward.

donny-dont commented on vcpkg.json:
```diff
@@ -0,0 +1,92 @@
+          ]
+        }
+      ]
+    },
```
@fujii please confirm these are the features you are expecting for harfbuzz.

donny-dont reviewed (commented):
Expanded the use of `vcpkg.json` so we can use it for feature selection.

fujii commented on Tools/Scripts/webkitdirs.pm:
```diff
@@ -2805,7 +2806,10 @@ sub generateBuildSystemFromCMakeProject
-    if (isPlayStation()) {
+    if (isAnyWindows()) {
+        push @args, '-DCMAKE_TOOLCHAIN_FILE="' . $ENV{VCPKG_ROOT} . '\\scripts\\buildsystems\\vcpkg.cmake"';
+        push @args, '-DVCPKG_TARGET_TRIPLET=x64-windows-webkit'
```
Why do you have to create a new triplet `x64-windows-webkit`? Use built-in triplet `x64-windows`.

fujii replied:
`vcpkg help triplet` for more info.

donny-dont replied:
Its required to replace `zlib` with `zlib-ng` see https://github.com/microsoft/vcpkg/pull/43147

fujii commented on Tools/Scripts/webkitdirs.pm:
```diff
@@ -2946,6 +2950,27 @@ sub cmakeArgsFromFeatures(\@;$)
     return @args;
 }
 
+sub vcpkgArgsFromFeatures(\@;$)
```
vcpkgArgsFromFeatures is useless. It just returns a constant string.
I think we don't have to add avif and skia now. It's easy to add them later.

donny-dont replied:
I haven't gotten around to looking for feature flags yet so this is stubbed out.

fujii replied:
Then, let's remove it for now.

donny-dont commented on Tools/Scripts/webkitdirs.pm:
```diff
@@ -2805,7 +2806,10 @@ sub generateBuildSystemFromCMakeProject
 
-    if (isPlayStation()) {
+    if (isAnyWindows()) {
+        push @args, '-DCMAKE_TOOLCHAIN_FILE="' . $ENV{VCPKG_ROOT} . '\\scripts\\buildsystems\\vcpkg.cmake"';
```
Should probably add an error if `VCPKG_ROOT` is not set.

foopoiuyt commented on Tools/Scripts/build-webkit:
```diff
@@ -224,10 +222,14 @@ if (!isCMakeBuild()) {
     die "Error: No WebKitLibraries directory found. Please do a fresh checkout.\n";
 }
 
+if (isAnyWindows() && ! defined $ENV{"VCPKG_ROOT"}) {
```
isAnyWindows seems to be "is running on windows or cygwin as it checks $^O, not doing the windows port (which I think is isWin).  This seems wrong since I think we only want to be using vcpkg for the one port and jsconly on windows.

foopoiuyt commented on Tools/Scripts/download-github-release.py:
```diff
@@ -1,162 +0,0 @@
-#! /usr/bin/env python3
```
This looks to be used downstream, so should have that resolved before removing the file.

donny-dont replied:
Pretty sure we stopped using GitHub releases internally so this script is of no use to it.

foopoiuyt commented on Tools/Scripts/webkitdirs.pm:
```diff
@@ -2805,7 +2806,10 @@ sub generateBuildSystemFromCMakeProject
     push @args, "-DLTO_MODE=$ltoMode" if ltoMode();
 
-    if (isPlayStation()) {
+    if (isAnyWindows()) {
```
As above, this seems like a bad choice of condition as it appears to be based on what OS perl was built for. We presumably want to do this for the windows port or jsconly on windows but not for another port that happens to be being built on windows.

foopoiuyt commented on Tools/Scripts/webkitdirs.pm:
```diff
@@ -2946,6 +2950,47 @@ sub cmakeArgsFromFeatures(\@;$)
+    my ($featuresArrayRef, $enableExperimentalFeatures) = @_;
+
+    my @args;
+    my $avif = 0;
```
Are these intended to be the default values? It seems unfortunate to encode a default in an additional place.

donny-dont replied:
Yep there really isnt any way around that without parsing the CMake files.

donny-dont commented on WebKitLibraries/triplets/x64-windows-webkit.cmake:
```diff
@@ -0,0 +1,54 @@
+set(VCPKG_BUILD_TYPE release)
+set(_BAD_PORT "curl")
+set(_BAD_VERSION "8.11.1")
+message(WARNING "Overriding VCPKG_BUILD_TYPE to ${VCPKG_BUILD_TYPE} due to an issue in ${_BAD_PORT} v${_BAD_VERSION}")
```
This will force all ports to be built with release only.

I tried having just curl specified this way but when it got to WebKit it used all release libraries and didn't mix and match so might as well optimize for this case.

donny-dont commented on WebKitLibraries/triplets/x64-windows-webkit.cmake:
```diff
@@ -0,0 +1,54 @@
+        )
+        message(FATAL_ERROR "Comment me out until you determine if the update fixes the issue")
+    endif ()
+endif ()
```
This is a sanity check to see if the version was modified. If so it directs the person making the modification to check the debug build and make a judgement call on whether or not this entire thing can be removed.

donny-dont reviewed (commented):
~~This forces release libraries for the moment until the curl issue is resolved. There could be other asserts being hit in other libraries that will not be caught at this time because of `VCPKG_BUILD_TYPE` being set.~~

This doesn't work due to libwoff2 requiring static linkage. So removing comments.

fujii reviewed (changes requested):
`build-webkit --debug` fails. Please test debug build&test before updating this PR.

I'm maintaining 14 Windows buildbot workers. Downloading tarball from official servers will make them flaky.
We should use [Binary Caching](https://learn.microsoft.com/en-us/vcpkg/users/binarycaching).

donny-dont committed:
[CMake] Integrate vcpkg into Windows build
https://bugs.webkit.org/show_bug.cgi?id=282800

Reviewed by Ross Kirsling.

Use `vcpkg` to handle the third party requirements of the Windows WebKit build.
The requirements will be built locally within the build and no longer
distributed from WebKitForWindows/WebKitRequirements. While all port definitions
are in that repository a migration to the canonical vcpkg registry at
Microsoft/vcpkg will be prioritized.

Removes the code that would download prebuilt binaries to `WebKitLibraries/win`.

This change will be used to support additional libraries in the future.

* .gitignore:
* Source/cmake/OptionsJSCOnly.cmake:
* Source/cmake/OptionsWin.cmake:
* Tools/Scripts/build-webkit:
* Tools/Scripts/download-github-release.py: Removed.
* Tools/Scripts/update-webkit-win-libs.py:
* Tools/Scripts/webkitdirs.pm:
(shouldUseVcpkg):
(generateBuildSystemFromCMakeProject):
(vcpkgArgsFromFeatures):
(windowsLibrariesDir): Deleted.
* Tools/Scripts/webkitperl/BuildSubproject.pm:
* WebKitLibraries/triplets/x64-windows-webkit.cmake: Added.
* vcpkg-configuration.json: Added.
* vcpkg.json: Added.

Canonical link: https://commits.webkit.org/295042@main
